core/planner.go

```diff
@@ -0,0 +1,138 @@
+	if a[i].Price == a[j].Price {
+		return a[i].Start.After(a[j].Start)
+	}
+	return a[i].Price < a[j].Price
+}
+
+func (a RatesByPrice) Swap(i, j int) {
+	a[i], a[j] = a[j], a[i]
+}
+
+type Planner struct {
+	log    *util.Logger
+	clock  clock.Clock // mockable time
+	tariff api.Tariff
+}
+
+func NewPlanner(log *util.Logger, tariff api.Tariff) *Planner {
+	clock := clock.New()
+	return &Planner{
+		log:    log,
+		clock:  clock,
+		tariff: tariff,
+	}
+}
+
+func (t *Planner) PlanActive(requiredDuration time.Duration, targetTime time.Time) (bool, error) {
+	if targetTime.Before(t.clock.Now()) || requiredDuration <= 0 {
+		return false, nil
+	}
+
+	data, err := t.tariff.Rates()
+	if err != nil {
+		return false, err
+	}
+
+	// rates are by default sorted by date, oldest to newest
+	last := data[len(data)-1].End
+
+	// sort rates by price
+	sort.Sort(RatesByPrice(data))
+
+	requiredDuration = time.Duration(float64(requiredDuration) / chargeEfficiency)
+
+	// Save same duration until next price info update
+	if targetTime.After(last) {
+		duration_old := requiredDuration
+		requiredDuration = time.Duration(float64(requiredDuration) * float64(time.Until(last)) / float64(time.Until(targetTime)))
+		t.log.DEBUG.Printf("reduced duration from %s to %s until got new priceinfo after %s\n", duration_old.Round(time.Minute), requiredDuration.Round(time.Minute), last.Round(time.Minute))
+	}
+
+	t.log.DEBUG.Printf("charge duration: %s, end: %v, find best prices:\n", requiredDuration.Round(time.Minute), targetTime.Round(time.Minute))
+
+	var cheapSlotNow bool
+	var cntExpectedSlots, curSlotNr int
+	var plannedDuration time.Duration
+
+	for _, slot := range data {
+		// timeslot not relevant
+		if slot.Start.After(targetTime) || slot.End.Before(t.clock.Now()) {
+			continue
+		}
+
+		// current slot
+		if slot.Start.Before(t.clock.Now()) && slot.End.After(t.clock.Now()) {
+			slot.Start = t.clock.Now()
+		}
+
+		// slot ends after target time
+		if slot.End.After(targetTime) {
+			slot.End = targetTime
+		}
+
+		cntExpectedSlots++
+		plannedDuration += slot.End.Sub(slot.Start)
+		t.log.TRACE.Printf("  Slot from: %v to %v price %f, timesum %s",
+			slot.Start.Round(time.Second), slot.End.Round(time.Second),
+			slot.Price, plannedDuration)
+
+		// current timeslot is a cheap one
+		if slot.Start.Before(t.clock.Now()) && slot.End.After(t.clock.Now()) {
+			cheapSlotNow = true
+			curSlotNr = cntExpectedSlots
+			t.log.TRACE.Printf(" (now, slot number %v)", curSlotNr)
+		}
+
+		// we found all necessary cheap slots to charge to targetSoC
+		if plannedDuration > requiredDuration {
+			break
+		}
+	}
+
+	var cheapactive bool
+	if cheapSlotNow {
+		if curSlotNr == cntExpectedSlots { // delay most expensive slot if not last slot
+			if cntExpectedSlots == 1 { // last slot
+				t.log.DEBUG.Printf("continue charging in last slot\n")
+				cheapactive = true
+			} else { // expensiv and not last slot, delay
+				if plannedDuration > requiredDuration+hysteresisDuration {
```

Comment:
Warum brauchen wir die Hysterese nochmal zusätzlich zur Effizienz? Ist das nicht ein bisschen viel Overkill? Ich wäre dafür die rauszuschmeissen und das _nur_ über Effizienz abzubilden.

Reply:
ein zusätzlicher Puffer war eigentlich nicht der Grund für die Hysterese. Ich hatte Probleme dass mehrfach ein/ausgeschaltet wurde wenn sich der SoC des Autos gesprungen ist. Damals war es auch wirklich eine Hysterese, diesen Zweck erfüllt es aktuell so nicht mehr. Ich muss das in echt ohne dieser Hysterese nochmals testen um zu sehen ob es noch einen negativen Effekt hat wenn ich es weglasse.

Reply:
@schenlap würdest Du da nochmal rein schauen? Vllt. einen Testcase für mit/ohne Hysterese bauen?